[MRG] early decode bytes returned by check_output

ogrisel opened the pull request:
Decode bytes strings into unicode strings as soon as we get the output of a subprocess command.

Let's wait for travis to check that this is working properly.

ogrisel commented:
Before merging I would like to add another test.

arjoly commented on clusterlib/tests/test_scheduler.py:
```diff
@@ -133,9 +133,7 @@ def test_queued_or_running_jobs_slurm():
-            if sys.version_info[0] == 3:
-                line = line.decode("utf8")
-            name, id_ = line.split()
+            name, id_ = line.decode("utf8").split()
```
Do we still the decode here?

ogrisel replied:
yes, every time we call `check_output`.

arjoly replied:
Thanks !

arjoly commented on clusterlib/scheduler.py:
```diff
@@ -101,14 +101,15 @@ def _slurm_queued_or_running_jobs(user=None):
 
 
-def queued_or_running_jobs(user=None):
+def queued_or_running_jobs(user=None, encoding='utf-8'):
```
Can you add a docstring for the new parameter and explain the use case to the user?

arjoly commented:
Nice addition !

ogrisel commented:
The new test uses non-ascii UTF-8 encoded chars. If both SLURM and SGE accept it I will document the encoding parameter. If not I will refactor the tests to:

- for ASCII encoding
- check that non-ASCII job name fail with a `ValueError` with a meaningful message.

ogrisel commented:
@arjoly I have updated the docstest to have them pass.

The `submit` function now always return a utf-8 encoded by string. That might be confusing for some users but it makes it always safe to call `os.system` on that whatever the version of Python.

Alternatevely, I could change that so that the `submit` function always return a unicode string. This would probably feel more intutive in most cases. If the user put non-ascii chars in the `job_name` parameter that would be his/her responsability to encode the output of the `submit` to utf-8 prior to passing it to `os.system` or `subprocess.call`. WDYT?

ogrisel commented:
BTW, both SGE and SLURM seem to accept UTF-8 encoded, non-ascii characters in the job name.

arjoly commented:
> The submit function now always return a utf-8 encoded by string. That might be confusing for some users but it makes it always safe to call os.system on that whatever the version of Python.

Why it's confusing?
I like this option. For me, it seems to be the user friendly choice and always gives stable output.
In which extent is the backward compatibility preserved?

(strings have still a part of mystery for me).

ogrisel commented:
> Why it's confusing?

Because users might expect to do (unicode) string operations (such as calling print on it) on them that don't work well in Python 3 without first decoding. For instance to print a byte string for visualization you have to decode it first.

In the current state of this branch, that return encoded byte string you get the following behavior with Python 3:

```
>>> print(submit(command="my_program", job_name="test_job"))
b'echo \'#!/bin/bash\nmy_program\' | qsub -N "test_job" -l h_rt=24:00:00 -l h_vmem=4000M'
>>> print(submit("my_program", job_name="test_job", backend='sge').decode('utf-8'))
echo '#!/bin/bash
my_program' | qsub -N "test_job" -l h_rt=24:00:00 -l h_vmem=4000M
```

Python 2 does it implicitly for you but it can be problematic if the platform specifc encoding used by Python 2 is not the one that was used to encode the string in the first place.

> For me, it seems to be the user friendly choice and always gives stable output.

If we return unicode string (that is Python 2 `unicode` or Python 3 `str`), they are still stable.

> In which extent is the backward compatibility preserved?

If we make the `submit` method return the unicode string you this will be backward compatible

ogrisel commented:
I pushed a new commit with that change. If you like the current state. I will add a new comment to add the docstring for the `encoding` parameter of the `queued_or_running_jobs`.

ogrisel commented:
Alright I added the doc. I also updated the users guide to give some insight on how to properly encode the command generated by `submit` when the `job_name` parameters contains non-ASCII chars. I can squash the commits if you wish.

arjoly commented:
Before a final review, do you also prefer this solution over the unicode one?

ogrisel commented:
> Before a final review, do you also prefer this solution over the unicode one?

I prefer the current state of this PR.

arjoly commented:
Can you remove file.txt?

arjoly commented:
Can you highlight your contribution in the what's new?

arjoly commented:
Beside the above comment, LGTM!

ogrisel commented:
Comments addressed. This should be good to merge. Sorry for the leftover `file.txt` file.

arjoly commented:
Merge by rebase !!!

Thanks @ogrisel for improving clusterlib!!!
